src/libcmd/installables.cc

```diff
@@ -521,33 +521,65 @@ InstallableFlake::InstallableFlake(
         throw UsageError("'--arg' and '--argstr' are incompatible with flakes");
 }
 
+namespace {
+std::optional<Path> getSourceFilePathFromInStorePath(Store & store, const Path & file, const fetchers::Input & flakeInput) {
+    if (!store.isInStore(file)) {
+        return std::nullopt;
+    }
+
+    auto [sourceStorePath, relPath] = store.toStorePath(file);
+    auto realPath = store.toRealPath(sourceStorePath);
+
+    auto maybeFlakeSource = flakeInput.getSourcePath();
+    if (!maybeFlakeSource) {
+        return std::nullopt;
+    }
+
+    auto betterFilePath = replaceStrings(file, realPath, *maybeFlakeSource);
+    if (!pathExists(betterFilePath)) {
+        return std::nullopt;
+    }
+
+    return betterFilePath;
+}
+}
+
 std::tuple<std::string, FlakeRef, InstallableValue::DerivationInfo> InstallableFlake::toDerivation()
 {
-    auto lockedFlake = getLockedFlake();
+    try {
```

Comment:
Maybe factor out another method to perform all the logic inside the `try…catch` block? The function is starting to get a bit hairy.

Actually the `try...catch` block itself could possibly be wrapped into its own top-level function (if that’s not too ugly) because it’ll be useful for the other entrypoints to the evaluation